compiler/rustc_query_system/src/values.rs

```diff
@@ -0,0 +1,16 @@
+use rustc_errors::FatalError;
+
+use crate::dep_graph::DepContext;
+
+pub trait Value<CTX: DepContext>: Sized {
+    fn from_cycle_error(tcx: CTX) -> Self;
+}
+
+impl<CTX: DepContext, T> Value<CTX> for T {
+    default fn from_cycle_error(tcx: CTX) -> T {
+        tcx.sess().abort_if_errors();
+        // Ideally we would use `bug!("called without errors")` here. But bug! is only defined in rustc_middle,
+        // and it's non-trivial to define it earlier. Raise a fatal error instead, which is hopefully close enough.
+        FatalError.raise()
```

Comment:
A `panic!` would be fine, it tends to be used as a replacement for `bug` when not available.
r=me either way.